File: modules/confgen/tests/test_confgen.c
/*
 * Copyright (c) 2018 Balabit
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "apphook.h"
#include "cfg-grammar.h"
#include "mock-cfg-parser.h"
#include <criterion/criterion.h>

#define TESTDATA_DIR TOP_SRCDIR "/modules/confgen/tests"

CfgParserMock *parser = NULL;

static void
_input(const gchar *input)
{
  cfg_parser_mock_input(parser, input);
}

static void
_next_token(void)
{
  cfg_parser_mock_next_token(parser);
}

static YYSTYPE *
_current_token(void)
{
  return parser->yylval;
}

#define assert_token_type(expected)                                     \
  cr_assert_eq(_current_token()->type, expected, "Unexpected token type %d != %d", _current_token()->type, expected);

#define assert_parser_string(expected)                          \
  _next_token();                                                        \
  assert_token_type(LL_STRING);                                        \
  cr_assert_str_eq(_current_token()->cptr, expected, "Unexpected string value parsed >>>%s<<< != >>>%s<<<", _current_token()->cptr, expected);

#define assert_parser_identifier(expected) \
  _next_token();                                                        \
  assert_token_type(LL_IDENTIFIER);                                         \
  cr_assert_str_eq(_current_token()->cptr, expected, "Unexpected identifier parsed >>>%s<<< != >>>%s<<<", _current_token()->cptr, expected);


Test(confgen, confgen_script_output_is_included_into_the_config)
{
  parser->lexer->ignore_pragma = FALSE;

  cfg_lexer_push_context(parser->lexer, main_parser.context, main_parser.keywords, main_parser.name);
  _input(
    "@module confgen context(root) name(confgentest) exec("TESTDATA_DIR "/confgentest.sh)\n"
    "from-config1\n"
    "confgentest()\n"
    "from-config2\n");

  assert_parser_identifier("from-config1");
  assert_parser_identifier("from-confgen1");
  assert_parser_identifier("from-confgen2");
  assert_parser_identifier("from-config2");
  cfg_lexer_pop_context(parser->lexer);
}

static void
setup(void)
{
  app_startup();
  configuration = cfg_new_snippet();
  parser = cfg_parser_mock_new();
}

static void
teardown(void)
{
  cfg_parser_mock_free(parser);
  cfg_free(configuration);
  configuration = NULL;
  app_shutdown();
}

TestSuite(confgen, .init = setup, .fini = teardown);

File: modules/python-modules/syslogng/modules/hypr/tests/test_confgen.py
#############################################################################
# Copyright (c) 2022 Balazs Scheidler <bazsi77@gmail.com>
# Copyright (c) 2023 Attila Szakacs
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# As an additional exemption you are allowed to compile & link against the
# OpenSSL libraries as published by the OpenSSL project. See the file
# COPYING for details.
#
#############################################################################
from syslogng.modules import hypr
import base64


def test_config_generator_all_options(mocker):
    mocker.patch("requests.get", return_value=mocker.Mock(**{
        'status_code': 200,
        'json': mocker.Mock(return_value=[
            {"appID": "HYPRDefaultApplication"},
            {"appID": "HYPRDefaultWorkstationApplication"},
            {"appID": "rp_foo"},
        ])
    }))

    bearer_token = b'bearer_token'
    base64_encoded_bearer_token = base64.b64encode(bearer_token).decode('utf8')
    hypr_args = {
        'url': '"https://dummy.hypr.com/"',
        'bearer_token': '"' + base64_encoded_bearer_token + '"',
    }

    snippet = hypr._hypr_config_generator(hypr_args)
    expected_snippet = """
        python-fetcher(
            class("syslogng.modules.hypr.HyprAuditSource")
            options(
                "url" => "https://dummy.hypr.com/"
                "rp_app_id" => "rp_foo"
                "bearer_token" => "%s"
                "page_size" => 100
                "initial_hours" => 4
                "ignore_persistence" => False
                "log_level" => "INFO"
            )
            flags()
            persist-name("hypr-https://dummy.hypr.com/-rp_foo")
            fetch-no-data-delay(60)
        );""" % base64_encoded_bearer_token

    assert expected_snippet in snippet


def test_config_generator_all_options(mocker):
    mocker.patch("requests.get", return_value=mocker.Mock(**{
        'status_code': 200,
        'json': mocker.Mock(return_value=[{"appID": "rp_foo"}, {"appID": "rp_bar"}, {"appID": "rp_baz"}])
    }))

    bearer_token = b'bearer_token'
    base64_encoded_bearer_token = base64.b64encode(bearer_token).decode('utf8')
    hypr_args = {
        'url': '"https://dummy.hypr.com/"',
        'bearer_token': '"' + base64_encoded_bearer_token + '"',
        'page_size': "123",
        'initial_hours': "9",
        'log_level': 'ERROR',
        'application_skiplist': '"rp_baz", "rp_app"',
        'ignore_persistence': "yes",
        'persist_name': '"custom-persist-name"',
        'fetch_no_data_delay': "42",
        'flags': 'no-parse',
    }

    snippet = hypr._hypr_config_generator(hypr_args)
    expected_snippets = [
        """
        python-fetcher(
            class("syslogng.modules.hypr.HyprAuditSource")
            options(
                "url" => "https://dummy.hypr.com/"
                "rp_app_id" => "rp_foo"
                "bearer_token" => "%s"
                "page_size" => 123
                "initial_hours" => 9
                "ignore_persistence" => yes
                "log_level" => "ERROR"
            )
            flags(no-parse)
            persist-name("custom-persist-name-rp_foo")
            fetch-no-data-delay(42)
        );""" % base64_encoded_bearer_token,
        """
        python-fetcher(
            class("syslogng.modules.hypr.HyprAuditSource")
            options(
                "url" => "https://dummy.hypr.com/"
                "rp_app_id" => "rp_bar"
                "bearer_token" => "%s"
                "page_size" => 123
                "initial_hours" => 9
                "ignore_persistence" => yes
                "log_level" => "ERROR"
            )
            flags(no-parse)
            persist-name("custom-persist-name-rp_bar")
            fetch-no-data-delay(42)
        );""" % base64_encoded_bearer_token,
    ]

    for expected_snippet in expected_snippets:
        assert expected_snippet in snippet
